### PR TITLE
ngroup support for SpatialFullConvolution

### PR DESCRIPTION
**Pull request by qiuxin2012**
## What changes were proposed in this pull request?

ngroup support for SpatialFullConvolution

## How was this patch tested?

unit tests
integration tests and manual tests (by @SeaOfOcean )

### CODE REVIEW

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 1]
```diff
@@ -150,19 +171,19 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
 
     require(kW > 0 && kH > 0, s"kernel size should be greater than zero, but got kH: $kH kW: $kW")
     require(dW > 0 && dH > 0, s"stride should be greater than zero, but got dH: $dH dW: $dW")
-    require(weight.nDimension == 2 || weight.nDimension == 4,
-      s"2D or 4D weight tensor expected, but got size: ${weight.size()}")
+    require(weight.nDimension == 3 || weight.nDimension == 5,
+      s"2D or 4D weight tensor expected, but got size: ${weight.dim()}")
```
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: not 2D or 4D, but 3D or 5D

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 2]
```diff
@@ -150,19 +171,19 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
...
 
     if (null != bias) {
-      require(bias.nDimension() == 1 && bias.size(1) == weight.size(2))
+      require(bias.nDimension() == 1 && bias.size(1) == weight.size(3) * weight.size(1))
```
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: add require comment

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 3]
```diff
@@ -219,94 +296,64 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
 
     // Resize output
     output.resize(batchSize, nOutputPlane, outputHeight, outputWidth)
+    output.zero()
 
-    // Resize temporary columns
-    if(null == columns) {
-      columns = Tensor[T]()
+    if (onesBias.dim() != 1 || onesBias.size(1) != outputHeight * outputWidth) {
+      onesBias.resize(Array(outputHeight * outputWidth)).fill(ev.fromType(1.0))
```
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: ev.one

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 4]
```diff
@@ -219,94 +296,64 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
...
-          bias.storage().array(), bias.storageOffset() - 1, k,
-          ev.fromType[Int](1),
-          output_n.storage().array(), output_n.storageOffset() - 1, n
-        )
-      }
+      val _i = elt
+      results(_i - 1) = Engine.model.invoke(() => {
+        // Matrix mulitply per output:
+        val input_n = inputTensor.select(1, _i)
+        require(input_n.isContiguous())
```
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: require comment

File: dl/src/test/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolutionSpec.scala [hunk 5]
```diff
@@ -64,8 +66,18 @@ class SpatialFullConvolutionSpec extends FlatSpec with Matchers {
...
   }
+
+  "A SpatialFullConvolution BilinearFiller" should "generate correct output" in {
+    val conv = new SpatialFullConvolution[Tensor[Double], Double](10, 10, 4, 4, 2, 2,
+      1, 1, 0, 0, 10, false, BilinearFiller)
+
+    val input = Tensor[Double](1, 10, 20, 30).apply1(e => Random.nextDouble())
+    val output = conv.updateOutput(input).toTensor[Double]
+
+    println(output)
```
Comment (zhangxiaoli73) [inline dl/src/test/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolutionSpec.scala]: how to verify output is correct?

File: dl/src/test/scala/com/intel/analytics/bigdl/torch/SpatialFullConvolutionSpec.scala [hunk 6]
```diff
@@ -350,11 +351,69 @@ class SpatialFullConvolutionSpec extends FlatSpec with BeforeAndAfter with Match
...
+    val nOutputPlane = 6
+    val kW = 3
+    val kH = 3
+    val dW = 1
+    val dH = 1
+    val layer = new SpatialFullConvolution[Tensor[Double], Double](nInputPlane, nOutputPlane,
+      kW, kH, dW, dH, 0, 0, 0, 0, 6)
+    Random.setSeed(3)
+    val input = Tensor[Double](6, 5, 5).apply1(e => Random.nextDouble())
+    layer.forward(input).toTensor[Double]
```
Comment (zhangxiaoli73) [inline dl/src/test/scala/com/intel/analytics/bigdl/torch/SpatialFullConvolutionSpec.scala]: what is it?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 7]
```diff
@@ -150,19 +171,19 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
...
+      require(bias.nDimension() == 1 && bias.size(1) == weight.size(3) * weight.size(1))
     }
 
     val ndim = input.nDimension
     val dimf = if (ndim == 4) 2 else 1
     val dimh = if (ndim == 4) 3 else 2
     val dimw = if (ndim == 4) 4 else 3
 
-    require(ndim == 3 || ndim == 4, s"3D or 4D input tensor expected but got size: ${input.size()}")
+    require(ndim == 3 || ndim == 4, s"3D or 4D input tensor expected but got size: ${input.dim()}")
```
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: add more info for requirement, such as  layer and function name.

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 8]
```diff
@@ -183,6 +204,63 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
...
+      ev.fromType[Int](1),
+      input.storage().array(), input.storageOffset() - 1, n,
+      weight.storage().array(), weight.storageOffset() - 1, m,
+      ev.fromType[Int](0),
+      columns.storage().array(), columns.storageOffset() - 1, n
+    )
+
+    if (!_1x1) {
+      val before = System.nanoTime()
+      ev.getType() match {
```
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: too many same getType() conditions, can you add a funciton to reuse?

File: dl/src/test/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolutionSpec.scala (hunk 5 above)
Comment (qiuxin2012) [inline dl/src/test/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolutionSpec.scala]: Tested in torch/SpatailFullConvolutionSpec, will delete this.

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 9]
```diff
@@ -457,83 +619,63 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
...
     if (ones.nDimension != 2 || ones.size(1) * ones.size(2) < outputHeight * outputWidth) {
       // Resize plane and fill with ones...
       ones.resize(outputHeight, outputWidth)
       ones.fill(ev.fromType[Int](1))
     }
 
-    // Resize temporary columns
-    columns.resize(nOutputPlane * kW * kH, inputHeight * inputWidth)
+    if (onesBatch.dim() != 1 || onesBatch.size(1) != batchSize) {
+      onesBatch.resize(Array(batchSize)).fill(ev.fromType(1.0))
```
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: ev.one

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 10]
```diff
@@ -148,38 +169,109 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
...
+    // (see http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-gemm)
+    val m = weight.size(2)
+    val n = columns.size(2)
+    val k = weight.size(1)
+
+    // Do GEMM (note: this is a bit confusing because gemm assumes column-major matrices)
+    DenseTensorBLAS.gemm[T](
+      'N', 'T',
+      n, m, k,
+      ev.fromType[Int](1),
```
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: ev.one

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 11]
```diff
@@ -148,38 +169,109 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
...
+      input.storage().array(), input.storageOffset() - 1, n,
+      weight.storage().array(), weight.storageOffset() - 1, m,
+      ev.fromType[Int](0),
```
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: ev.zero

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 12]
```diff
@@ -317,6 +378,57 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
...
+      case FloatType => NNPrimitive.im2colWithDilationFloat(
+        gradOutput.asInstanceOf[Tensor[Float]], columns.asInstanceOf[Tensor[Float]],
+        gradOutput.size(1), outputHeight, outputWidth,
+        kH, kW,
+        padH, padW,
+        dH, dW,
+        1, 1
+      )
+
+      case _ => throw new UnsupportedOperationException(s"Only Float/Double supported")
```
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: more clear comment

**Pull request merged**